whisper/src/message.rs

```diff
@@ -0,0 +1,476 @@
+	Decoder(DecoderError),
+	LivesTooLong,
+	IssuedInFuture,
+	ZeroTTL,
+}
+
+impl From<DecoderError> for Error {
+	fn from(err: DecoderError) -> Self {
+		Error::Decoder(err)
+	}
+}
+
+impl fmt::Display for Error {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		match *self {
+			Error::Decoder(ref err) => write!(f, "Failed to decode message: {}", err),
+			Error::LivesTooLong => write!(f, "Message claims to be issued before the unix epoch."),
+			Error::IssuedInFuture => write!(f, "Message issued in future."),
+			Error::ZeroTTL => write!(f, "Message live for zero time."),
+		}
+	}
+}
+
+// Raw envelope struct.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Envelope {
+	/// Expiry timestamp
+	pub expiry: u64,
+	/// Time-to-live in seconds
+	pub ttl: u64,
+	/// series of 4-byte topics.
+	pub topics: SmallVec<[Topic; 4]>,
+	/// The message contained within.
+	pub data: Vec<u8>,
+	/// Arbitrary value used to target lower PoW hash.
+	pub nonce: u64,
+}
+
+impl Envelope {
+	fn proving_hash(&self) -> H256 {
+		use byteorder::{BigEndian, ByteOrder};
+
+		let mut buf = [0; 32];
+
+		let mut stream = RlpStream::new_list(4);
+		stream.append(&self.expiry)
+			.append(&self.ttl)
+			.append_list(&self.topics)
+			.append(&self.data);
+
+		let mut digest = Keccak::new_keccak256();
+		digest.update(&*stream.drain());
+		digest.update(&{
+			let mut nonce_bytes = [0u8; 8];
+			BigEndian::write_u64(&mut nonce_bytes, self.nonce);
+
+			nonce_bytes
+		});
+
+		digest.finalize(&mut buf);
+		H256(buf)
+	}
+}
+
+impl rlp::Encodable for Envelope {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		s.begin_list(5)
+			.append(&self.expiry)
+			.append(&self.ttl)
+			.append_list(&self.topics)
+			.append(&self.data)
+			.append(&self.nonce);
+	}
+}
+
+impl rlp::Decodable for Envelope {
+	fn decode(rlp: &UntrustedRlp) -> Result<Self, DecoderError> {
+		if rlp.item_count()? != 5 { return Err(DecoderError::RlpIncorrectListLen) }
+
+		Ok(Envelope {
+			expiry: rlp.val_at(0)?,
+			ttl: rlp.val_at(1)?,
+			topics: rlp.at(2)?.iter().map(|x| x.as_val()).collect::<Result<_, _>>()?,
+			data: rlp.val_at(3)?,
+			nonce: rlp.val_at(4)?,
+		})
+	}
+}
+
+/// Message creation parameters.
+/// Pass this to `Message::create` to make a message.
+pub struct CreateParams {
+	/// time-to-live in seconds.
+	pub ttl: u64,
+	/// payload data.
+	pub payload: Vec<u8>,
+	/// Topics.
+	pub topics: Vec<Topic>,
+	/// How many milliseconds to spend proving work.
+	pub work: u64,
+}
+
+/// A whisper message. This is a checked message carrying around metadata.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Message {
+	envelope: Envelope,
+	bloom: H512,
+	hash: H256,
+	encoded_size: usize,
+}
+
+impl Message {
+	/// Create a message from creation parameters.
+	/// Panics if TTL is 0.
+	pub fn create(params: CreateParams) -> Self {
+		use byteorder::{BigEndian, ByteOrder};
+		use rand::{Rng, SeedableRng, XorShiftRng};
+
+		let mut rng = {
+			let mut thread_rng = ::rand::thread_rng();
```

Comment:
it's only used to find a nonce which brings the PoW below the threshold. Speed here would be the critical factor. From an outside observer's perspective it doesn't matter whether the method used to choose the nonce is high or low entropy. It could even be consecutive